state/protocol/protocol_state/statemachine.go

```diff
@@ -46,3 +46,86 @@ type ProtocolStateMachine interface {
+	nextEpochIdentitiesLookup    map[flow.Identifier]*flow.DynamicIdentityEntry // lookup for nodes active in the next epoch, may be nil or empty
+}
+
+// Build returns updated protocol state entry, state ID and a flag indicating if there were any changes.
+func (u *baseProtocolStateMachine) Build() (updatedState *flow.ProtocolStateEntry, stateID flow.Identifier, hasChanges bool) {
+	updatedState = u.state.Copy()
+	stateID = updatedState.ID()
+	hasChanges = stateID != u.parentState.ID()
+	return
+}
+
+// View returns the view that is associated with this state updater.
+// The view of the StateUpdater equals the view of the block carrying the respective updates.
+func (u *baseProtocolStateMachine) View() uint64 {
+	return u.view
+}
+
+// ParentState returns parent protocol state that is associated with this state updater.
+func (u *baseProtocolStateMachine) ParentState() *flow.RichProtocolStateEntry {
+	return u.parentState
+}
+
+// ensureLookupPopulated ensures that current and next epoch identities lookups are populated.
+// We use this to avoid populating lookups on every UpdateIdentity call.
+func (u *baseProtocolStateMachine) ensureLookupPopulated() {
+	if len(u.currentEpochIdentitiesLookup) > 0 {
+		return
+	}
+	u.rebuildIdentityLookup()
+}
+
+// rebuildIdentityLookup re-generates lookups of *active* participants for
+// previous (optional, if u.state.PreviousEpoch ≠ nil), current (required) and
+// next epoch (optional, if u.state.NextEpoch ≠ nil).
+func (u *baseProtocolStateMachine) rebuildIdentityLookup() {
+	if u.state.PreviousEpoch != nil {
+		u.prevEpochIdentitiesLookup = u.state.PreviousEpoch.ActiveIdentities.Lookup()
+	} else {
+		u.prevEpochIdentitiesLookup = nil
+	}
+	u.currentEpochIdentitiesLookup = u.state.CurrentEpoch.ActiveIdentities.Lookup()
+	if u.state.NextEpoch != nil {
+		u.nextEpochIdentitiesLookup = u.state.NextEpoch.ActiveIdentities.Lookup()
+	} else {
+		u.nextEpochIdentitiesLookup = nil
+	}
+}
+
+// UpdateIdentity updates identity table with new identity entry.
+// Should pass identity which is already present in the table, otherwise an exception will be raised.
+// No errors are expected during normal operations.
+func (u *baseProtocolStateMachine) UpdateIdentity(updated *flow.DynamicIdentityEntry) error {
+	u.ensureLookupPopulated()
+	prevEpochIdentity, foundInPrev := u.prevEpochIdentitiesLookup[updated.NodeID]
+	if foundInPrev {
+		prevEpochIdentity.Dynamic = updated.Dynamic
+	}
+	currentEpochIdentity, foundInCurrent := u.currentEpochIdentitiesLookup[updated.NodeID]
+	if foundInCurrent {
+		currentEpochIdentity.Dynamic = updated.Dynamic
+	}
+	nextEpochIdentity, foundInNext := u.nextEpochIdentitiesLookup[updated.NodeID]
+	if foundInNext {
+		nextEpochIdentity.Dynamic = updated.Dynamic
+	}
+	if !foundInPrev && !foundInCurrent && !foundInNext {
+		return fmt.Errorf("expected to find identity for prev, current or next epoch, but (%v) was not found", updated.NodeID)
+	}
```

Comment:
We should adjust the error handling here to ensure it does not cause the chain to halt (or cause Epoch Fallback Mode to be triggered). At the moment, I believe this would halt the chain, which we certainly don't want. Until we have a way to recover from EFM without a spork, we should prioritize liveness and discard these invalid events (with a hefty and noticeable warning or alert).

Reply:
I have updated `UpdateIdentity` to return a sentinel error, once we introduce a dedicated service event for ejection we will just log an error instead of entering EFM. I believe for this PR nothing more can be done.